x-pack/plugins/viz_editor/public/components/main/main.tsx

```diff
@@ -84,6 +106,19 @@ export function Main({ kfetch }: Props) {
     return <h1>Loading...</h1>;
   }
 
+  const { editorPanels, toExpression } = registry.getByName(state.currentEditorConfig);
+
+  const { leftPanel, rightPanel } = editorPanels({
+    standardState: state.standardVisState,
+    customState: state.customVisState,
+    onChangeStandardState: (newState: StandardVisState) => {
```

Comment:
I'm not totally convinced that there's a difference between "standard" and "custom". I think that we just have state, and it's pluggable. So, a plugin (probably the core) adds in "title" and "xAxis" and "yAxis" and "query" and other things that we consider to be "standard", and another plugin might add in "zAxis" and a third plugin might rely on "zAxis", etc.

Not sure. And as for how a plugin updates state, I'm not sure. We could do essentially what you're doing here, which is pass a `setState` or `onChangeState` or whatever, down into the plugins, and they can literally just call that to change any portion of the state. It feels like this would be a bit brittle and make re-arranging state tricky, but honestly, this is probably the most direct and obvious approach.

Alternatively, we force everyone to read state via selectors and update state via updaters / reducers / whatever, so that we can move bits of state around, and as long as the selectors / updaters are correct, the plugins relying on that state won't break.

Reply:
What I’ve been trying to fight with this private state was the brittleness you mentioned - the public standard state is well defined for all plugins and should be treated like a public API - the private state is a scratchpad for that plugin which it can use to do everything, and it’s guaranteed to not break any other editor. Selectors + Updates are another layer on top of the interface for the public state, but I’m not sure whether we need that layer of abstraction in the beginning.

Reply:
It is basically this part from your doc: 

> - If a visualization requires new state which is probably not going to be shared by any other plugin, it should register its own root extension (e.g. `kibana_lineChart`, `x-pack_scatterPlot`).

You are absolutely right that most stuff should go into the standard state to not be locked in into one editor config. Maybe we should call this `privateState` and `publicState` - I'm not sure.